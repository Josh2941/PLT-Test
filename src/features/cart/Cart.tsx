import React, { useEffect, useState } from "react"
import { FlatList, Text, View } from "react-native";
import { ListItem, ListItemProps } from "../../components/ListItem";
import { connect } from "react-redux";
import * as baseActions from '../../redux/BaseActions';

type ListProps = {
    selectedList: ListItemProps[]
    actions: {
        increment: (item: ListItemProps) => void,
        decrement: (item: ListItemProps) => void
    }
}

export const Cart = ({ actions: { increment, decrement }, selectedList }: ListProps) => {

    const [ cartList, setCartList ] = useState<ListItemProps[]>(selectedList);

    useEffect(() => {
        if (cartList !== selectedList) 
            setCartList(selectedList);
    }, [selectedList])

    return (
        <View style={{ flex: 1 }}>
            <FlatList
                testID="cartList"
                style={{ flex: 1 }}
                data={cartList}
                extraData={cartList}
                renderItem={({ item }) => <ListItem {...item} onPressIncrement={increment} onPressDecrement={decrement} />}/>
            <Text>Total: <Text testID="total">{selectedList.reduce((acc, {price, selectedCount}) => acc + (price * selectedCount!), 0)}</Text></Text>
        </View>
    )
}

const mapStateToProps = (state: any) => {
    return {
        selectedList: state.base.items
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        actions: {
            increment: (item: ListItemProps) => dispatch(baseActions.increment(item)),
            decrement: (item: ListItemProps) => dispatch(baseActions.decrement(item))
        }
    }
}

const CartConnected = connect(mapStateToProps, mapDispatchToProps)(Cart);
const CartContainer = (props: any) => <CartConnected {...props} />;

export default CartContainer