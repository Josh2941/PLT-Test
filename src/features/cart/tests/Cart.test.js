import renderer, { act } from 'react-test-renderer'
import { Cart } from '../Cart'

const mockOnPressIncrement = jest.fn();
const mockOnPressDecrement = jest.fn();

const MOCK_PROPS = {
    selectedList: [
        {
          "id": 1,
          "colour": "Black",
          "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
          "price": 10,
          "selectedCount": 3,
          "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024"
        },
        {
          "id": 2,
          "colour": "Stone",
          "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
          "price": 4,
          "selectedCount": 2,
          "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024"
        }
    ],
    actions: {
        increment: mockOnPressIncrement,
        decrement: mockOnPressDecrement
    }
}

describe('Test Cart', () => {
    test('renders correctly', () => {
        const tree = renderer.create(<Cart {...MOCK_PROPS} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('show cart list correctly', () => {
        const component = renderer.create(<Cart {...MOCK_PROPS} />).root;
        const cartList = component.findByProps({ testID: 'cartList' });
        expect(cartList.props.data.length).toBeGreaterThan(0);
    });
    test('show cart list total correctly', () => {
        const component = renderer.create(<Cart {...MOCK_PROPS} />).root;
        const cartTotal = component.findByProps({ testID: 'total' });
        expect(cartTotal.props.children).toBe(MOCK_PROPS.selectedList.reduce((acc, {price, selectedCount}) => acc + (price * selectedCount), 0));
    });
})
