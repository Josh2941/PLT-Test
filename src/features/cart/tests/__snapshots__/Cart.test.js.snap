// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cart renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      [
        {
          "colour": "Black",
          "id": 1,
          "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
          "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
          "price": 10,
          "selectedCount": 3,
        },
        {
          "colour": "Stone",
          "id": 2,
          "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
          "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
          "price": 4,
          "selectedCount": 2,
        },
      ]
    }
    extraData={
      [
        {
          "colour": "Black",
          "id": 1,
          "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
          "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
          "price": 10,
          "selectedCount": 3,
        },
        {
          "colour": "Stone",
          "id": 2,
          "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
          "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
          "price": 4,
          "selectedCount": 2,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    testID="cartList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flex": 1,
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <Image
            source={
              {
                "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
              }
            }
            style={
              {
                "height": 100,
                "width": 100,
              }
            }
          />
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <Text
              style={
                {
                  "marginBottom": 10,
                  "width": "80%",
                }
              }
              testID="listItemName"
            >
              Black Sheet Strappy Textured Glitter Bodycon Dress
            </Text>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View>
                <Text
                  style={
                    {
                      "marginBottom": 10,
                    }
                  }
                  testID="listItemColor"
                >
                  Black
                </Text>
                <Text
                  style={
                    {
                      "marginBottom": 10,
                    }
                  }
                  testID="listItemPrice"
                >
                  10
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="listItemDecermentButton"
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#007AFF",
                            "fontSize": 18,
                            "margin": 8,
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      -
                    </Text>
                  </View>
                </View>
                <Text
                  testID="listItemCount"
                >
                  3
                </Text>
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="listItemIncermentButton"
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#007AFF",
                            "fontSize": 18,
                            "margin": 8,
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      +
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flex": 1,
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
              }
            }
            style={
              {
                "height": 100,
                "width": 100,
              }
            }
          />
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <Text
              style={
                {
                  "marginBottom": 10,
                  "width": "80%",
                }
              }
              testID="listItemName"
            >
              Stone Ribbed Strappy Cut Out Detail Bodycon Dress
            </Text>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View>
                <Text
                  style={
                    {
                      "marginBottom": 10,
                    }
                  }
                  testID="listItemColor"
                >
                  Stone
                </Text>
                <Text
                  style={
                    {
                      "marginBottom": 10,
                    }
                  }
                  testID="listItemPrice"
                >
                  4
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="listItemDecermentButton"
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#007AFF",
                            "fontSize": 18,
                            "margin": 8,
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      -
                    </Text>
                  </View>
                </View>
                <Text
                  testID="listItemCount"
                >
                  2
                </Text>
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="listItemIncermentButton"
                >
                  <View
                    style={
                      [
                        {},
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#007AFF",
                            "fontSize": 18,
                            "margin": 8,
                            "textAlign": "center",
                          },
                        ]
                      }
                    >
                      +
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Text>
    Total: 
    <Text
      testID="total"
    >
      38
    </Text>
  </Text>
</View>
`;
